Safety Stock=(Maximum Daily Sales×Maximum Lead Time)−(Average Daily Sales×Average Lead Time)
Reorder Point=(Average Daily Sales×Average Lead Time)+Safety Stock

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
sales_df = pd.read_csv("../data/processed/clean_sales.csv")
sales_df.head()

,id,item_id,dept_id,cat_id,store_id,state_id,d_1,d_2,d_3,d_4,...,d_1904,d_1905,d_1906,d_1907,d_1908,d_1909,d_1910,d_1911,d_1912,d_1913
0,HOBBIES_1_001_CA_1_validation,HOBBIES_1_001,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,...,1,3,0,1,1,1,3,0,1,1
1,HOBBIES_1_002_CA_1_validation,HOBBIES_1_002,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,...,0,0,0,0,0,1,0,0,0,0
2,HOBBIES_1_003_CA_1_validation,HOBBIES_1_003,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,...,2,1,2,1,1,1,0,1,1,1
3,HOBBIES_1_004_CA_1_validation,HOBBIES_1_004,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,...,1,0,5,4,1,0,1,3,7,2
4,HOBBIES_1_005_CA_1_validation,HOBBIES_1_005,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,...,2,1,1,0,1,1,2,2,2,4


In [4]:
daily_sales_columns = [
    col for col in sales_df.columns
    if col.startswith("d_")
]

print("Number of daily sales columns:", len(daily_sales_columns))
print("First 10 columns:", daily_sales_columns[:10])
print("Last 10 columns:", daily_sales_columns[-10:])

Number of daily sales columns: 1913
First 10 columns: ['d_1', 'd_2', 'd_3', 'd_4', 'd_5', 'd_6', 'd_7', 'd_8', 'd_9', 'd_10']
Last 10 columns: ['d_1904', 'd_1905', 'd_1906', 'd_1907', 'd_1908', 'd_1909', 'd_1910', 'd_1911', 'd_1912', 'd_1913']


In [5]:
sales_df["average_daily_sales"] = sales_df[
    daily_sales_columns
].mean(axis=1)

sales_df[
    ["item_id", "store_id", "average_daily_sales"]
].head(10)

,item_id,store_id,average_daily_sales
0,HOBBIES_1_001,CA_1,0.313643
1,HOBBIES_1_002,CA_1,0.257710
2,HOBBIES_1_003,CA_1,0.150549
3,HOBBIES_1_004,CA_1,1.718766
4,HOBBIES_1_005,CA_1,0.966545
5,HOBBIES_1_006,CA_1,0.858338
6,HOBBIES_1_007,CA_1,0.221641
7,HOBBIES_1_008,CA_1,7.229482
8,HOBBIES_1_009,CA_1,1.186095
9,HOBBIES_1_010,CA_1,0.719289


In [6]:
sales_df["maximum_daily_sales"] = sales_df[
    daily_sales_columns
].max(axis=1)

sales_df[
    ["item_id", "store_id", "maximum_daily_sales"]
].head(10)

,item_id,store_id,maximum_daily_sales
0,HOBBIES_1_001,CA_1,5
1,HOBBIES_1_002,CA_1,5
2,HOBBIES_1_003,CA_1,6
3,HOBBIES_1_004,CA_1,15
4,HOBBIES_1_005,CA_1,9
5,HOBBIES_1_006,CA_1,10
6,HOBBIES_1_007,CA_1,3
7,HOBBIES_1_008,CA_1,91
8,HOBBIES_1_009,CA_1,20
9,HOBBIES_1_010,CA_1,6


In [7]:
AVERAGE_LEAD_TIME = 7
MAXIMUM_LEAD_TIME = 14

sales_df["average_lead_time"] = AVERAGE_LEAD_TIME
sales_df["maximum_lead_time"] = MAXIMUM_LEAD_TIME

print("Average Lead Time:", AVERAGE_LEAD_TIME, "days")
print("Maximum Lead Time:", MAXIMUM_LEAD_TIME, "days")

Average Lead Time: 7 days
Maximum Lead Time: 14 days


Safety stock

For HOBBIES_1_001:

Average Daily Sales ≈ 0.3136
Maximum Daily Sales = 5
Average Lead Time = 7
Maximum Lead Time = 14

Safety Stock
= (5 × 14) - (0.3136 × 7)
= 70 - 2.1952
≈ 67.80 units

In [8]:
sales_df["safety_stock"] = (
    (sales_df["maximum_daily_sales"] * sales_df["maximum_lead_time"])
    -
    (sales_df["average_daily_sales"] * sales_df["average_lead_time"])
)

sales_df[
    [
        "item_id",
        "store_id",
        "average_daily_sales",
        "maximum_daily_sales",
        "average_lead_time",
        "maximum_lead_time",
        "safety_stock"
    ]
].head(10)

,item_id,store_id,average_daily_sales,maximum_daily_sales,average_lead_time,maximum_lead_time,safety_stock
0,HOBBIES_1_001,CA_1,0.313643,5,7,14,67.804496
1,HOBBIES_1_002,CA_1,0.257710,5,7,14,68.196027
2,HOBBIES_1_003,CA_1,0.150549,6,7,14,82.946158
3,HOBBIES_1_004,CA_1,1.718766,15,7,14,197.968636
4,HOBBIES_1_005,CA_1,0.966545,9,7,14,119.234187
5,HOBBIES_1_006,CA_1,0.858338,10,7,14,133.991636
6,HOBBIES_1_007,CA_1,0.221641,3,7,14,40.448510
7,HOBBIES_1_008,CA_1,7.229482,91,7,14,1223.393623
8,HOBBIES_1_009,CA_1,1.186095,20,7,14,271.697334
9,HOBBIES_1_010,CA_1,0.719289,6,7,14,78.964976


Reorder Point=(Average Daily Sales×Average Lead Time)+Safety Stock

For HOBBIES_1_001:

Average Daily Sales ≈ 0.3136
Average Lead Time = 7
Safety Stock ≈ 67.80

Reorder Point
= (0.3136 × 7) + 67.80
≈ 70.00 units

In [9]:
sales_df["reorder_point"] = (
    (sales_df["average_daily_sales"] * sales_df["average_lead_time"])
    + sales_df["safety_stock"]
)

sales_df[
    [
        "item_id",
        "store_id",
        "average_daily_sales",
        "average_lead_time",
        "safety_stock",
        "reorder_point"
    ]
].head(10)

,item_id,store_id,average_daily_sales,average_lead_time,safety_stock,reorder_point
0,HOBBIES_1_001,CA_1,0.313643,7,67.804496,70.0
1,HOBBIES_1_002,CA_1,0.257710,7,68.196027,70.0
2,HOBBIES_1_003,CA_1,0.150549,7,82.946158,84.0
3,HOBBIES_1_004,CA_1,1.718766,7,197.968636,210.0
4,HOBBIES_1_005,CA_1,0.966545,7,119.234187,126.0
5,HOBBIES_1_006,CA_1,0.858338,7,133.991636,140.0
6,HOBBIES_1_007,CA_1,0.221641,7,40.448510,42.0
7,HOBBIES_1_008,CA_1,7.229482,7,1223.393623,1274.0
8,HOBBIES_1_009,CA_1,1.186095,7,271.697334,280.0
9,HOBBIES_1_010,CA_1,0.719289,7,78.964976,84.0


In [10]:
inventory_optimization = sales_df[
    [
        "item_id",
        "store_id",
        "average_daily_sales",
        "maximum_daily_sales",
        "average_lead_time",
        "maximum_lead_time",
        "safety_stock",
        "reorder_point"
    ]
].copy()

inventory_optimization.head(10)

,item_id,store_id,average_daily_sales,maximum_daily_sales,average_lead_time,maximum_lead_time,safety_stock,reorder_point
0,HOBBIES_1_001,CA_1,0.313643,5,7,14,67.804496,70.0
1,HOBBIES_1_002,CA_1,0.257710,5,7,14,68.196027,70.0
2,HOBBIES_1_003,CA_1,0.150549,6,7,14,82.946158,84.0
3,HOBBIES_1_004,CA_1,1.718766,15,7,14,197.968636,210.0
4,HOBBIES_1_005,CA_1,0.966545,9,7,14,119.234187,126.0
5,HOBBIES_1_006,CA_1,0.858338,10,7,14,133.991636,140.0
6,HOBBIES_1_007,CA_1,0.221641,3,7,14,40.448510,42.0
7,HOBBIES_1_008,CA_1,7.229482,91,7,14,1223.393623,1274.0
8,HOBBIES_1_009,CA_1,1.186095,20,7,14,271.697334,280.0
9,HOBBIES_1_010,CA_1,0.719289,6,7,14,78.964976,84.0


In [11]:
numeric_columns = [
    "average_daily_sales",
    "maximum_daily_sales",
    "average_lead_time",
    "maximum_lead_time",
    "safety_stock",
    "reorder_point"
]

inventory_optimization[numeric_columns] = (
    inventory_optimization[numeric_columns].round(2)
)

inventory_optimization.head(10)

,item_id,store_id,average_daily_sales,maximum_daily_sales,average_lead_time,maximum_lead_time,safety_stock,reorder_point
0,HOBBIES_1_001,CA_1,0.31,5,7,14,67.80,70.0
1,HOBBIES_1_002,CA_1,0.26,5,7,14,68.20,70.0
2,HOBBIES_1_003,CA_1,0.15,6,7,14,82.95,84.0
3,HOBBIES_1_004,CA_1,1.72,15,7,14,197.97,210.0
4,HOBBIES_1_005,CA_1,0.97,9,7,14,119.23,126.0
5,HOBBIES_1_006,CA_1,0.86,10,7,14,133.99,140.0
6,HOBBIES_1_007,CA_1,0.22,3,7,14,40.45,42.0
7,HOBBIES_1_008,CA_1,7.23,91,7,14,1223.39,1274.0
8,HOBBIES_1_009,CA_1,1.19,20,7,14,271.70,280.0
9,HOBBIES_1_010,CA_1,0.72,6,7,14,78.96,84.0


In [12]:
inventory_optimization.to_csv(
    "../data/processed/inventory_optimization.csv",
    index=False
)

print("Inventory Optimization file saved successfully!")
print("Shape:", inventory_optimization.shape)

Inventory Optimization file saved successfully!
Shape: (30490, 8)


In [13]:
import os

output_path = "../data/processed/inventory_optimization.csv"

print("File exists:", os.path.exists(output_path))

File exists: True


In [14]:
inventory_optimization.head(10)

,item_id,store_id,average_daily_sales,maximum_daily_sales,average_lead_time,maximum_lead_time,safety_stock,reorder_point
0,HOBBIES_1_001,CA_1,0.31,5,7,14,67.80,70.0
1,HOBBIES_1_002,CA_1,0.26,5,7,14,68.20,70.0
2,HOBBIES_1_003,CA_1,0.15,6,7,14,82.95,84.0
3,HOBBIES_1_004,CA_1,1.72,15,7,14,197.97,210.0
4,HOBBIES_1_005,CA_1,0.97,9,7,14,119.23,126.0
5,HOBBIES_1_006,CA_1,0.86,10,7,14,133.99,140.0
6,HOBBIES_1_007,CA_1,0.22,3,7,14,40.45,42.0
7,HOBBIES_1_008,CA_1,7.23,91,7,14,1223.39,1274.0
8,HOBBIES_1_009,CA_1,1.19,20,7,14,271.70,280.0
9,HOBBIES_1_010,CA_1,0.72,6,7,14,78.96,84.0
